src/v2.rs

```diff
@@ -0,0 +1,957 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::future::Future;
+use std::mem::size_of;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::Poll;
+use std::task::Waker;
+
+use deno_ast::EmitOptions;
+use deno_ast::TranspiledSource;
+use deno_graph::ModuleGraph;
+use futures::future::poll_fn;
+use sha2::Digest;
+use sha2::Sha256;
+use tokio::io::AsyncReadExt;
+use url::Url;
+
+use crate::error::ParseError;
+use crate::Module;
+use crate::ModuleInner;
+use crate::ModuleKind;
+
+pub(crate) const ESZIP_V2_MAGIC: &[u8; 8] = b"ESZIP_V2";
+
+#[derive(Debug, PartialEq)]
+#[repr(u8)]
+enum HeaderFrameKind {
+  Module = 0,
+  Redirect = 1,
+}
+
+/// Version 2 of the Eszip format. This format supports streaming sources and
+/// source maps.
+#[derive(Debug, Default)]
+pub struct EszipV2 {
+  modules: Arc<Mutex<HashMap<String, EszipV2Module>>>,
+  pub ordered_modules: Vec<String>,
+}
+
+#[derive(Debug)]
+enum EszipV2Module {
+  Module {
+    kind: ModuleKind,
+    source: EszipV2SourceSlot,
+    source_map: EszipV2SourceSlot,
+  },
+  Redirect {
+    target: String,
+  },
+}
+
+#[derive(Debug)]
+enum EszipV2SourceSlot {
+  Pending {
+    offset: usize,
+    length: usize,
```

Comment:
```suggestion
    offset: u32,
    length: u32,
```
Should be `u32` to reflect the parser logic

Reply:
Later code needs this to be a usize though, as slice index needs to be done with `usize`s. It is easier to convert once, than at each site of use.